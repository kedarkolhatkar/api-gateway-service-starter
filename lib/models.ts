import { RestApi, JsonSchemaType, Model, RequestValidator } from '@aws-cdk/aws-apigateway';

export interface UserServiceModels {
  createUserRequestModel: Model;
  createUserSuccessResponseModel: Model;
  internalErrorResponseModel: Model;
  getUserResponseModel: Model;
}

export interface UserServiceValidators {
  bodyOnlyValidator: RequestValidator;
}

export class ModelBuilder {
  private restApi: RestApi;

  constructor(restApi: RestApi) {
    this.restApi = restApi;
  }

  buildModels(): UserServiceModels {
    const createUserRequestModel = this.restApi.addModel('create-user-request-model-id', {
      modelName: 'createUserRequestModel',
      schema: {
        type: JsonSchemaType.OBJECT,
        properties: {
          username: {
            type: JsonSchemaType.STRING,
          },
          firstName: {
            type: JsonSchemaType.STRING,
          },
          lastName: {
            type: JsonSchemaType.STRING,
          },
          primaryEmail: {
            type: JsonSchemaType.STRING,
            format: 'email',
          },
          primaryPhone: {
            type: JsonSchemaType.STRING,
            pattern: '^d{3}-d{2}-d{4}$',
          },
        },
        required: ['username', 'firstName', 'lastName'],
      },
      contentType: 'application/json',
      description: 'Model for request object for createUser API method',
    });

    const createUserSuccessResponseModel = this.restApi.addModel('create-user-response-model-id', {
      modelName: 'createUserSuccessResponseModel',
      schema: {
        type: JsonSchemaType.OBJECT,
        properties: {
          id: {
            type: JsonSchemaType.STRING,
            format: 'uuid',
          },
        },
        required: ['id'],
      },
      contentType: 'application/json',
      description: 'Model for response object for createUser API method',
    });

    const internalErrorResponseModel = this.restApi.addModel('inter-error-response-model-id', {
      modelName: 'internalErrorResponseModel',
      schema: {
        type: JsonSchemaType.OBJECT,
        properties: {
          errorCode: {
            type: JsonSchemaType.STRING,
          },
          message: {
            type: JsonSchemaType.STRING,
          },
        },
      },
    });

    const getUserResponseModel = this.restApi.addModel('get-user-response-model-id', {
      modelName: 'getUserResponseModel',
      schema: {
        type: JsonSchemaType.OBJECT,
        properties: {
          id: {
            type: JsonSchemaType.STRING,
            format: 'uuid',
          },
          username: {
            type: JsonSchemaType.STRING,
          },
          firstName: {
            type: JsonSchemaType.STRING,
          },
          lastName: {
            type: JsonSchemaType.STRING,
          },
          primaryEmail: {
            type: JsonSchemaType.STRING,
            format: 'email',
          },
          primaryPhone: {
            type: JsonSchemaType.STRING,
            pattern: '^d{3}-d{2}-d{4}$',
          },
        },
        required: ['id', 'username', 'firstName', 'lastName'],
      },
      contentType: 'application/json',
      description: 'Model for response object for getUser API method',
    });

    return {
      createUserRequestModel,
      createUserSuccessResponseModel,
      internalErrorResponseModel,
      getUserResponseModel,
    };
  }

  buildValidators(): UserServiceValidators {
    const bodyOnlyValidator = this.restApi.addRequestValidator('bodyOnlyRequestValidator', {
      requestValidatorName: 'bodyOnlyValidator',
      validateRequestBody: true,
      validateRequestParameters: false,
    });

    return {
      bodyOnlyValidator,
    };
  }
}
