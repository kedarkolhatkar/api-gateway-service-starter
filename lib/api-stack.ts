import * as cdk from '@aws-cdk/core';
import * as lambda from '@aws-cdk/aws-lambda';
import * as path from 'path';
import * as apigateway from '@aws-cdk/aws-apigateway';
import * as dynamodb from '@aws-cdk/aws-dynamodb';
import { TableEncryption } from '@aws-cdk/aws-dynamodb';
import { ModelBuilder } from './models';

/**
 *
 */
export interface APIGatewayStackProps extends cdk.StackProps {
  restApiName: string;
  apiDescription: string;
  srcDirectory: string;
  lambdaFunctionName: string;
}

export class APIGatewayStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props: APIGatewayStackProps) {
    super(scope, id, props);

    const userTableName = 'user-table';
    const userTable = new dynamodb.Table(this, `${userTableName}-id`, {
      tableName: userTableName,
      partitionKey: { name: 'userId', type: dynamodb.AttributeType.STRING },
      billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
      encryption: TableEncryption.AWS_MANAGED,
    });

    const lambdaFunctionSrcDir = `${props.srcDirectory}/${props.lambdaFunctionName}`;
    // console.log(path.join(__dirname, lambdaFunctionSrcDir));

    // Define the backend lambda
    const backend = new lambda.Function(this, `${props.lambdaFunctionName}-id`, {
      functionName: props.lambdaFunctionName,
      runtime: lambda.Runtime.NODEJS_14_X,
      handler: `${props.lambdaFunctionName}.handler`,
      code: lambda.Code.fromAsset(path.join(__dirname, lambdaFunctionSrcDir)),
      environment: {
        USER_TABLE_ARN: userTable.tableArn,
        USER_TABLE_NAME: userTableName,
      },
    });

    userTable.grantReadWriteData(backend);

    // Create User API
    const api = new apigateway.LambdaRestApi(this, 'user-api', {
      restApiName: props.restApiName,
      description: props.apiDescription,
      handler: backend,
      proxy: false,
    });

    const modelBuilder = new ModelBuilder(api);
    const models = modelBuilder.buildModels();
    const requestValidators = modelBuilder.buildValidators();

    // Build resources and methods
    const users = api.root.addResource('users');

    users.addMethod('PUT', undefined, {
      operationName: 'Create User',
      requestModels: {
        'application/json': models.createUserRequestModel,
      },
      methodResponses: [
        {
          statusCode: '201',
          responseModels: {
            'application/json': models.createUserResponseModel,
          },
        },
      ],
      requestValidator: requestValidators.bodyOnlyValidator,
    });

    const user = users.addResource('{user}');
    user.addMethod('GET', undefined, {
      operationName: 'Get user by Id',
    });
  }
}
