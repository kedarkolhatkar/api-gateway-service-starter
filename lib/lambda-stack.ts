import * as cdk from '@aws-cdk/core';
import * as lambda from '@aws-cdk/aws-lambda';
import * as path from 'path';

export class LambdaStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const srcDirectory = '../build';
    const lambdaFunctionName = 'user-service-handler';
    const lambdaFunctionSrcDir = `${srcDirectory}/${lambdaFunctionName}`;
    console.log(path.join(__dirname, lambdaFunctionSrcDir));

    // The code that defines your stack goes here
    const fn = new lambda.Function(this, `${lambdaFunctionName}-id`, {
      functionName: lambdaFunctionName,
      runtime: lambda.Runtime.NODEJS_14_X,
      handler: `${lambdaFunctionName}.handler`,
      code: lambda.Code.fromAsset(path.join(__dirname, lambdaFunctionSrcDir)),
    });
  }
}
