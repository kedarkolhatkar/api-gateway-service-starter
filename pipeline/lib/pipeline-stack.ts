import * as cdk from '@aws-cdk/core';
import * as codebuild from '@aws-cdk/aws-codebuild';
import * as codepipeline from '@aws-cdk/aws-codepipeline';
import * as codepipelineActions from '@aws-cdk/aws-codepipeline-actions';
import { LogGroup } from '@aws-cdk/aws-logs';
import { Effect, PolicyStatement, Role } from '@aws-cdk/aws-iam';

export interface PipelineStackProps extends cdk.StackProps {
  projectName: string;
  serviceName: string;
  buildspecsDir?: string;
  repoName: string;
  repoOwner: string;
  repoAccessTokenName: string;
  branch?: string;
}

export class PipelineStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props: PipelineStackProps) {
    super(scope, id, props);

    const defaultProps = { buildSpecsDir: 'pipeline/buildspecs', branch: 'main' };
    const finalProps = { ...defaultProps, ...props };

    // First build the code build project
    const cdkBuildProject = new codebuild.PipelineProject(this, `${finalProps.serviceName}-build-1`, {
      projectName: `${finalProps.serviceName}-build-project`,
      description: `This code build project for ${finalProps.serviceName} compiles, unit tests and deploys to development environment`,
      buildSpec: codebuild.BuildSpec.fromSourceFilename('pipeline/buildspecs/dev.yaml'),
      concurrentBuildLimit: 1,
      environment: {
        buildImage: codebuild.LinuxBuildImage.STANDARD_5_0,
      },
      logging: {
        cloudWatch: {
          enabled: true,
          logGroup: new LogGroup(this, `${finalProps.serviceName}-build-log-group-id`, {
            logGroupName: `${finalProps.serviceName}-build-logs`,
            retention: 7,
          }),
          prefix: `${finalProps.serviceName}`,
        },
      },
    });

    // Policy to create/update/delete AWS resources
    const awsAssetsPolicyStatement = new PolicyStatement({
      effect: Effect.ALLOW,
      resources: ['*'],
      actions: ['*'],
      conditions: {
        'ForAnyValue:StringEquals': {
          'aws:CalledVia': ['cloudformation.amazonaws.com'],
        },
      },
    });

    cdkBuildProject.addToRolePolicy(awsAssetsPolicyStatement);

    // Policy for access to CloudFormation stacks
    const accessStackPolicy = new PolicyStatement({
      effect: Effect.ALLOW,
      resources: ['arn:aws:cloudformation:us-east-1:729288039836:stack/*'],
      actions: [
        'cloudformation:DescribeStacks',
        'cloudformation:GetTemplate',
        'cloudformation:DeleteChangeSet',
        'cloudformation:CreateChangeSet',
        'cloudformation:DescribeChangeSet',
        'cloudformation:ExecuteChangeSet',
        'cloudformation:DescribeStackEvents',
      ],
    });
    cdkBuildProject.addToRolePolicy(accessStackPolicy);

    // Policy to access S3 bucket where deployment artificats are stored
    const s3bucketAccessPolicy = new PolicyStatement({
      effect: Effect.ALLOW,
      resources: ['arn:aws:s3:::cdk-*', 'arn:aws:s3:::codepipeline-us-east-1-*'],
      actions: [
        's3:PutObject',
        's3:GetObject',
        's3:GetObjectVersion',
        's3:GetBucketAcl',
        's3:GetBucketLocation',
        's3:ListBucket',
      ],
    });

    cdkBuildProject.addToRolePolicy(s3bucketAccessPolicy);

    const sourceOutput = new codepipeline.Artifact('SrcOutput');
    const githubToken = cdk.SecretValue.secretsManager(finalProps.repoAccessTokenName);

    // Create code pipeline
    // TODO: Use SecretManager to get token
    new codepipeline.Pipeline(this, `${finalProps.serviceName}-pipeline`, {
      restartExecutionOnUpdate: true,
      stages: [
        {
          stageName: 'Source',
          actions: [
            new codepipelineActions.GitHubSourceAction({
              actionName: 'Checkout',
              output: sourceOutput,
              repo: finalProps.repoName,
              branch: finalProps.branch,
              owner: finalProps.repoOwner,
              oauthToken: githubToken,
            }),
          ],
        },
        {
          stageName: 'Build',
          actions: [
            new codepipelineActions.CodeBuildAction({
              actionName: 'build',
              project: cdkBuildProject,
              input: sourceOutput,
            }),
          ],
        },
      ],
    });
  }
}
