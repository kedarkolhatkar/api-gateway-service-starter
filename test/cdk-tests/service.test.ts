import { expect as expectCDK, matchTemplate, MatchStyle, SynthUtils } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import { APIGatewayStack, APIGatewayStackProps } from '../../lib/api-stack';

test('APIGatewayStack', () => {
  const app = new cdk.App();
  // WHEN
  const props = {
    description: 'The API stack for user-service. This is a root level stack for the user microservice.',
    restApiName: 'userServiceAPI',
    apiDescription: 'This service provides functionality to manager users',
    srcDirectory: '../build',
    lambdaFunctionName: 'user-service-handler',
    terminationProtection: true,
  } as APIGatewayStackProps;

  const stack = new APIGatewayStack(app, 'MyTestStack', props);
  // THEN
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
