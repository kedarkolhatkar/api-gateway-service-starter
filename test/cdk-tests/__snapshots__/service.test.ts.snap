// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`APIGatewayStack 1`] = `
Object {
  "Description": "The API stack for user-service. This is a root level stack for the user microservice.",
  "Outputs": Object {
    "userapiEndpoint2656FFE6": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "userapiD4BE445D",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "userapiDeploymentStageprod950E1A85",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "AssetParameters16b93d6d261e367f70c0ba7cd022c8ab429ef316fabcf6a5e70930bc5a28090eArtifactHash164DA615": Object {
      "Description": "Artifact hash for asset \\"16b93d6d261e367f70c0ba7cd022c8ab429ef316fabcf6a5e70930bc5a28090e\\"",
      "Type": "String",
    },
    "AssetParameters16b93d6d261e367f70c0ba7cd022c8ab429ef316fabcf6a5e70930bc5a28090eS3Bucket7E75A85B": Object {
      "Description": "S3 bucket for asset \\"16b93d6d261e367f70c0ba7cd022c8ab429ef316fabcf6a5e70930bc5a28090e\\"",
      "Type": "String",
    },
    "AssetParameters16b93d6d261e367f70c0ba7cd022c8ab429ef316fabcf6a5e70930bc5a28090eS3VersionKey6DA6D977": Object {
      "Description": "S3 key for asset version \\"16b93d6d261e367f70c0ba7cd022c8ab429ef316fabcf6a5e70930bc5a28090e\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "userapiANY29A688AE": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "ANY",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "userservicehandleridB5F252C6",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Fn::GetAtt": Array [
            "userapiD4BE445D",
            "RootResourceId",
          ],
        },
        "RestApiId": Object {
          "Ref": "userapiD4BE445D",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "userapiANYApiPermissionMyTestStackuserapi222D21F1ANY0B3C3900": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "userservicehandleridB5F252C6",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "userapiD4BE445D",
              },
              "/",
              Object {
                "Ref": "userapiDeploymentStageprod950E1A85",
              },
              "/*/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "userapiANYApiPermissionTestMyTestStackuserapi222D21F1ANY55A988E9": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "userservicehandleridB5F252C6",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "userapiD4BE445D",
              },
              "/test-invoke-stage/*/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "userapiAccount42697875": Object {
      "DependsOn": Array [
        "userapiD4BE445D",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "userapiCloudWatchRoleAE5A4CA7",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
    },
    "userapiCloudWatchRoleAE5A4CA7": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "userapiD4BE445D": Object {
      "Properties": Object {
        "Description": "This service provides functionality to manager users",
        "Name": "userServiceAPI",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "userapiDeploymentC9949C0Fa17a2f6e2a2c29849c171c4616e426ee": Object {
      "DependsOn": Array [
        "userapiproxyANY20A17C16",
        "userapiproxy24E6BE40",
        "userapiANY29A688AE",
      ],
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "userapiD4BE445D",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "userapiDeploymentStageprod950E1A85": Object {
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "userapiDeploymentC9949C0Fa17a2f6e2a2c29849c171c4616e426ee",
        },
        "RestApiId": Object {
          "Ref": "userapiD4BE445D",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "userapiproxy24E6BE40": Object {
      "Properties": Object {
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "userapiD4BE445D",
            "RootResourceId",
          ],
        },
        "PathPart": "{proxy+}",
        "RestApiId": Object {
          "Ref": "userapiD4BE445D",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "userapiproxyANY20A17C16": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "ANY",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "userservicehandleridB5F252C6",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Ref": "userapiproxy24E6BE40",
        },
        "RestApiId": Object {
          "Ref": "userapiD4BE445D",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "userapiproxyANYApiPermissionMyTestStackuserapi222D21F1ANYproxy4907FD16": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "userservicehandleridB5F252C6",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "userapiD4BE445D",
              },
              "/",
              Object {
                "Ref": "userapiDeploymentStageprod950E1A85",
              },
              "/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "userapiproxyANYApiPermissionTestMyTestStackuserapi222D21F1ANYproxyDBBCDBC2": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "userservicehandleridB5F252C6",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "userapiD4BE445D",
              },
              "/test-invoke-stage/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "userservicehandleridB5F252C6": Object {
      "DependsOn": Array [
        "userservicehandleridServiceRoleAEF1BC0F",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters16b93d6d261e367f70c0ba7cd022c8ab429ef316fabcf6a5e70930bc5a28090eS3Bucket7E75A85B",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters16b93d6d261e367f70c0ba7cd022c8ab429ef316fabcf6a5e70930bc5a28090eS3VersionKey6DA6D977",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters16b93d6d261e367f70c0ba7cd022c8ab429ef316fabcf6a5e70930bc5a28090eS3VersionKey6DA6D977",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "FunctionName": "user-service-handler",
        "Handler": "user-service-handler.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "userservicehandleridServiceRoleAEF1BC0F",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "userservicehandleridServiceRoleAEF1BC0F": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
